File: Python/Test_pFN_Computation.py
# Yuncong Ma, 8/22/2023
# Test the computation of pFNs using precomputed group-level FNs

#########################################
# Packages
import nibabel as nib
import numpy as np
import scipy.io as sio
import os
import re
from scipy.io import savemat
from scipy.sparse import spdiags, lil_matrix
from scipy.spatial import distance
import time

import pNet

#########################################
# Example data from HCP surf Test_FN17
gFN = pNet.load_matlab_array(os.path.join(pNet.Example.HCP_surf.dir_pnet, 'Group_FN', 'FN.mat'), 'FN')
# gNb uses index starting from 1, but it starts from 0 in Python
gNb = pNet.load_matlab_single_array(os.path.join(pNet.Example.HCP_surf.dir_pnet, 'FN_Computation', 'gNb.mat'))
Data = pNet.load_matlab_single_array(os.path.join(pNet.Example.HCP_surf.dir_data, '100206', '1', 'LR', 'Image.mat'))
Data = Data.T
Setting = pNet.load_json_setting(os.path.join(pNet.Example.HCP_surf.dir_pnet, 'FN_Computation', 'Setting.json'))

#########################################
# Setup parameters
K = gFN.shape[1]
maxIter = 1000
minIter = 30
meanFitRatio = 0.1
error = 1e-4
normW = 1
eta = Setting['PersonalizedFN']['eta']
Alpha = 2
Beta = 30
alphaS = Setting['PersonalizedFN']['alphaS21']
alphaL = Setting['PersonalizedFN']['alphaL']
vxI = 1
initConv = 1
ard = Setting['PersonalizedFN']['ard']
dataPrecision = 'double'
logFile = 'Log_pFN_NMF.log'

# setup log file
logFile = open(logFile, 'w')
print(f'\nStart NMF for pFN using NumPy at '+time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(time.time()))+'\n', file=logFile)

#########################################
# Setup data precision and eps
np_float, np_eps = pNet.set_data_precision(dataPrecision)

Data = Data.astype(np_float)
gFN = gFN.astype(np_float)

# initialization
initV = gFN

dim_time, dim_space = Data.shape

# Median number of graph neighbors
nM = np.median(np.unique(gNb[:, 0], return_counts=True)[1])

# Use Alpha and Beta to set alphaS and alphaL if they are 0
if alphaS == 0 and Alpha > 0:
    alphaS = np.round(Alpha * dim_time / K)
if alphaL == 0 and Beta > 0:
    alphaL = np.round(Beta * dim_time / K / nM)

#########################################
# Prepare and normalize scan
X = pNet.normalize_data(Data.T, 'vp', 'vmax')
# savemat('pFN_X.mat', {'X': X})


# Construct the spatial affinity graph
L, W, D = pNet.construct_Laplacian_gNb(gNb, dim_space, vxI, X, alphaL, normW, dataPrecision)
# savemat('pFN_init_D.mat', {'D': D})
# savemat('pFN_init_L.mat', {'L': L})
# savemat('pFN_init_W_2.mat', {'W': W})


# Initialize V
V = np.copy(initV)
miv = np.max(V, axis=0)
trimInd = V / np.maximum(np.tile(miv, (dim_space, 1)), np_eps) < 5e-2
V[trimInd] = 0
# savemat('pFN_init_V.mat', {'V': V})

# Initialize U
U = X @ V / np.tile(np.sum(V, axis=0), (dim_time, 1))
# savemat('pFN_init_U_0.mat', {'U': U})

U = pNet.initialize_u(X, U, V, error, maxIter, minIter, meanFitRatio, initConv)
# savemat('pFN_init_U.mat', {'U': U})

initU = U.copy()
initV = V.copy()

# Alternative update of U and V
# Variables
U = initU.copy()
V = initV.copy()

if ard > 0:
    lambdas = np.sum(U, axis=0) / dim_time
    hyperLam = eta * np.sum(np.power(X, 2)) / (dim_time * dim_space * 2)
else:
    lambdas = 0
    hyperLam = 0

flagQC = 0
oldLogL = np.inf
oldU = U.copy()
oldV = V.copy()

#  Multiplicative update of U and V
for i in range(1, 1+maxIter):
    # ===================== update V ========================
    # Eq. 8-11
    XU = X.T @ U
    UU = U.T @ U
    VUU = V @ UU

    tmpl2 = np.power(V, 2)

    if alphaS > 0:
        tmpl21 = np.sqrt(tmpl2)
        tmpl22 = np.tile(np.sqrt(np.sum(tmpl2, axis=0)), (dim_space, 1))
        tmpl21s = np.tile(np.sum(tmpl21, axis=0), (dim_space, 1))
        posTerm = V / np.maximum(tmpl21 * tmpl22, np_eps)
        negTerm = V * tmpl21s / np.maximum(np.power(tmpl22, 3), np_eps)

        VUU = VUU + 0.5 * alphaS * posTerm
        XU = XU + 0.5 * alphaS * negTerm

    if alphaL > 0:
        WV = W @ V.astype(np.float64)
        DV = D @ V.astype(np.float64)

        XU = XU + WV
        VUU = VUU + DV

    V = V * (XU / np.maximum(VUU, np_eps))
    #savemat('pFN_V_1.mat', {'V': V})

    # Prune V if empty components are found in V
    # This is almost impossible to happen without combining FNs
    prunInd = np.sum(V != 0, axis=0) == 1
    if np.any(prunInd):
        V[:, prunInd] = np.zeros((dim_space, np.sum(prunInd)))
        U[:, prunInd] = np.zeros((dim_time, np.sum(prunInd)))

    # normalize U and V
    U, V = pNet.normalize_u_v(U, V, 1, 1)
    #savemat('pFN_U_2.mat', {'U': U})
    #savemat('pFN_V_2.mat', {'V': V})

    # ===================== update U =========================
    XV = X @ V
    VV = V.T @ V
    UVV = U @ VV

    if ard > 0:  # ard term for U
        posTerm = 1 / np.maximum(np.tile(lambdas, (dim_time, 1)), np_eps)

        UVV = UVV + posTerm * hyperLam

    U = U * (XV / np.maximum(UVV, np_eps))

    # Prune U if empty components are found in U
    # This is almost impossible to happen without combining FNs
    prunInd = np.sum(U, axis=0) == 0
    if np.any(prunInd):
        V[:, prunInd] = np.zeros((dim_space, np.sum(prunInd)))
        U[:, prunInd] = np.zeros((dim_time, np.sum(prunInd)))

    # update lambda
    if ard > 0:
        lambdas = np.sum(U, axis=0) / dim_time

    # ==== calculate objective function value ====
    ardU = 0
    tmp1 = 0
    tmp2 = 0
    tmp3 = 0
    tmpl21 = np.power(V, 2)

    if ard > 0:
        su = np.sum(U, axis=0)
        su[su == 0] = 1
        ardU = np.sum(np.log(su)) * dim_time * hyperLam

    if alphaL > 0:
        tmp2 = V.T @ L * V.T

    L21 = alphaS * np.sum(np.sum(np.sqrt(tmpl21), axis=0) / np.maximum(np.sqrt(np.sum(tmpl21, axis=0)), np_eps))
    LDf = pNet.data_fitting_error(X, U, V, 0, 1)
    LSl = np.sum(tmp2)

    # Objective function
    LogL = L21 + LDf + LSl + ardU
    print(f"    Iter = {i}: LogL: {LogL}, dataFit: {LDf}, spaLap: {LSl}, L21: {L21}, ardU: {ardU}", file=logFile)

    # The iteration needs to meet minimum iteration number and small changes of LogL
    if i > minIter and abs(oldLogL - LogL) / np.maximum(oldLogL, np_eps) < error:
        break
    oldLogL = LogL.copy()

    # QC Control
    temp = pNet.mat_corr(initV, V)
    QC_Spatial_Correspondence = np.copy(np.diag(temp))
    temp -= np.diag(np.diag(temp))
    QC_Spatial_Correspondence_Control = np.max(temp, axis=1)
    QC_Delta_Sim = np.min(QC_Spatial_Correspondence - QC_Spatial_Correspondence_Control)

    if QC_Delta_Sim <= 0:
        flagQC = 1
        U = oldU.copy()
        V = oldV.copy()
        print(f'\n  QC: Meet QC constraint: Delta sim = {QC_Delta_Sim}', file=logFile)
        print(f'    Use results from last iteration', file=logFile)
        break
    else:
        oldU = U.copy()
        oldV = V.copy()
        print(f'        QC: Delta sim = {QC_Delta_Sim}', file=logFile)

savemat('pFN_U_end.mat', {'U': U})
savemat('pFN_V_end.mat', {'V': V})

print(f'\n Finished at '+time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(time.time()))+'\n', file=logFile)

File: MATLAB/GUI/Private_Function/MATLAB_Yuncong/Development/Test_pFN_Computation.m
% Yuncong Ma, 8/14/2023
% This script is to test the corresponding python file, in order to ensure
% the same result of pFN computation
% Test_pFN_Computation.py

%% Add to path
addpath(genpath('/Users/yuncongma/Documents/Document/fMRI/Myworks/pNet'));
%%
maxNumCompThreads(10);
tic
% load file
scan = fLoad_MATLAB_Single_Variable('/Users/yuncongma/Documents/Document/fMRI/Myworks/pNet/Example/HCP_Surface/Data/100206/1/LR/Image.mat');
gNb = fLoad_MATLAB_Single_Variable('/Users/yuncongma/Documents/Document/fMRI/Myworks/pNet/Example/HCP_Surface/Test_FN17/FN_Computation/gNb.mat');
gFN = fLoad_MATLAB_Single_Variable('/Users/yuncongma/Documents/Document/fMRI/Myworks/pNet/Example/HCP_Surface/Test_FN17/Group_FN/FN.mat');
Setting = fLoad_MATLAB_Single_Variable('/Users/yuncongma/Documents/Document/fMRI/Myworks/pNet/Example/HCP_Surface/Test_FN17/FN_Computation/Setting.mat');

scan=single(scan);
gFN=single(gFN);
%%

% parameter
K = 17;
maxIter = 1000;
minIter = 30;
meanFitRatio = 0.1;
error = 1e-4;
nRepeat = 1;
rounds = 30;
normW = 1;
eta = 0;
alphaS = 2;
alphaL = 10;
vxI = 0;
initConv = 1;
ard = 0;
dataPrecision = 'double';
logFile = 'Log_pFN_NMF_MATLAB.log';

% setup log file
% fileID = fopen(logFile, 'w');
fprintf(['\nStart NMF for pFN using MATLAB at ',char(datetime),'\n'])


% [mat_float,mat_eps] = set_data_precision(dataPrecision);
mat_eps = eps(dataPrecision);
if strcmp(dataPrecision,'single')
    gFN=single(gFN);
    scan=single(scan);
else
    gFN=double(gFN);
    scan=double(scan);
end

% initialization
initV=gFN;

dim_space = size(scan,1);
dim_time = size(scan,2);

% Setup alphaS and alphaL
nmVec = zeros(length(gNb),1);
for gni=1:length(gNb)
    nmVec(gni) = length(gNb{gni});
end
nM = median(nmVec);

% Use Alpha and Beta to set alphaS and alphaL if they are 0
if alphaS==0 && Alpha>0
    alphaS = round((Alpha*dim_time)/K);
end
if alphaL==0 && Beta>0
    alphaL = round((Beta*dim_time)/(K*nM));
end

% normalize data
X = normalize_data(scan','vp','vmax');


% construct the spatial affinity graph
tmpW = sparse(dim_space,dim_space);
for vi=1:dim_space
    nei = gNb{vi};
    if vxI>0
        corrVal = (1+corr(X(:,vi),X(:,nei)))/2;
    else
        corrVal = 1;
    end
    tmpW(vi,nei) = corrVal;
    tmpW(nei,vi) = corrVal;
end
W = tmpW;


DCol = full(sum(W,2));
D = spdiags(DCol,0,dim_space,dim_space);
L = D - W;
if normW>0
    D_mhalf = spdiags(DCol.^-0.5,0,dim_space,dim_space);
    L = D_mhalf * L * D_mhalf * alphaL;
    W = D_mhalf * W * D_mhalf * alphaL;
    D = D_mhalf * D * D_mhalf * alphaL;
end

% initialize V
V = initV;
miv = max(V,[],1);
trimInd = V ./ max(repmat(miv,dim_space,1),mat_eps) < 5e-2;
V(trimInd)=0;


% initialize U
U = X * V ./ repmat(sum(V,1),[dim_time,1]);

U = initialize_u(X, U, V, error, maxIter, minIter, meanFitRatio, initConv);

initU=U;
initV=V;

%
% Alternative update of U and V

oldL = Inf;
j = 0;
iterLog = [];

% No reuse of initU and initV, can be reference
U = initU;
V = initV;

[dim_time,dim_space] = size(X);

if ard>0
    lambdas = sum(U,1) / dim_time;
    hyperLam = eta * sum(sum(X.^2,1),2) / (dim_time*dim_space*2);
else
    lambdas=0;
    hyperLam=0;
end

flagQC=0;
oldLogL=inf;
oldU=U;
oldV=V;
% Multiplicative update of U and V
for i=1:maxIter

    % ===================== update V ========================
    % Eq. 8-11
    XU = X'*U;
    UU = U'*U;
    VUU = V*UU;

    tmpl2 = V.^2;

    if alphaS>0
        tmpl21 = sqrt(tmpl2);
        tmpl22 = repmat(sqrt(sum(tmpl2,1)),dim_space,1);
        tmpl21s = repmat(sum(tmpl21,1),dim_space,1);
        posTerm = V ./ max(tmpl21.*tmpl22,mat_eps);
        negTerm = V .* tmpl21s ./ max(tmpl22.^3,mat_eps);

        VUU = VUU + 0.5 * alphaS * posTerm;
        XU = XU + 0.5 * alphaS * negTerm;
    end

    if alphaL>0
        WV = W * double(V);
        DV = D * double(V);

        XU = XU + WV;
        VUU = VUU + DV;
    end

    V = V.*(XU./max(VUU,mat_eps));

    prunInd = sum(V~=0,1)==1;
    if any(prunInd)
        V(:,prunInd) = zeros(dim_space,sum(prunInd));
        U(:,prunInd) = zeros(dim_time,sum(prunInd));
    end

    % ==== normalize U and V ====
    [U,V] = normalize_u_v(U, V, 1, 1);

    % ===================== update U =========================
    XV = X*V;
    VV = V'*V;
    UVV = U*VV;

    if ard>0 % ard term for U
        posTerm = 1./max(repmat(lambdas,dim_time,1),mat_eps);
        UVV = UVV + posTerm*hyperLam;
    end

    U = U.*(XV./max(UVV,mat_eps));

    prunInd = sum(U)==0;
    if any(prunInd)
        V(:,prunInd) = zeros(dim_space,sum(prunInd,1));
        U(:,prunInd) = zeros(dim_time,sum(prunInd,1));
    end

    % update lambda
    if ard>0
        lambdas = sum(U,1) / dim_time;
    end

    % ==== calculate partial objective function value ====
    ardU = 0;
    tmp1 = 0;
    tmp2 = 0;
    tmp3 = 0;
    tmpl21 = V.^2;

    if ard>0
        su=sum(U,1);
        su(su==0)=1;
        ardU=sum(log(su))*dim_time*hyperLam;
    end

    tmpDf=(X-U*V').^2;
    tmp1=sum(tmpDf(:));

    if alphaL>0
        tmp2=V'*L.*V';
    end

    L21=alphaS*sum(sum(sqrt(tmpl21),1)./max(sqrt(sum(tmpl21,1)),mat_eps));
    LDf = tmp1;
    LSl = sum(tmp2(:));

    % Objective function
    LogL = L21 + ardU + LDf + LSl;


    fprintf(['  Iter = ',num2str(i),': LogL:',num2str(LogL),',dataFit:',num2str(LDf)...
        ',spaLap:',num2str(LSl),...
        ',L21:',num2str(L21),',ardU:',num2str(ardU),'\n']);

    if i>minIter && abs(oldLogL-LogL)/max(oldLogL,mat_eps)<error
        break;
    end
    oldLogL = LogL;

    % QC Control
    temp=corr(initV,V);
    QC_Spatial_Correspondence=diag(temp);
    temp=temp-diag(diag(temp));
    QC_Spatial_Correspondence_Control=max(temp,[],2);
    QC_Delta_Sim=min(QC_Spatial_Correspondence-QC_Spatial_Correspondence_Control);
    if QC_Delta_Sim<=0
        U=oldU;
        V=oldV;
        flagQC=1;
        fprintf('\n  Meet QC constraint: Delta sim = %f\n',QC_Delta_Sim);
        fprintf('    Use results from last iteration\n');
        break;
    else
        oldU=U;
        oldV=V;
        fprintf('\n    Delta sim = %f\n',QC_Delta_Sim);
    end

end

fprintf(['\n Finished at ',char(datetime),'\n'])
% save();
toc


